File: agent/intelligence.py
import os
import json
import re

import vertexai
from vertexai.generative_models import GenerativeModel, GenerationConfig

from cache import insight_cache

vertexai.init(project=os.getenv("GOOGLE_CLOUD_PROJECT"), location="us-central1")

model = GenerativeModel("gemini-1.5-flash")
json_config = GenerationConfig(
    response_mime_type="application/json",
    temperature=0.3,
)


def parse_json(text):
    try:
        return json.loads(text)
    except json.JSONDecodeError:
        match = re.search(r"\{[\s\S]*\}", text)
        if match:
            try:
                return json.loads(match.group())
            except json.JSONDecodeError:
                pass
        match = re.search(r"\[[\s\S]*\]", text)
        if match:
            try:
                return json.loads(match.group())
            except json.JSONDecodeError:
                pass
    return None


def summarize_video(title: str, transcript: str, video_id: str = "") -> dict:
    cache_key = f"video:{video_id or title}"
    cached = insight_cache.get(cache_key)
    if cached:
        return cached

    prompt = f"""Given this YouTube video titled "{title}" with the following transcript, provide:
1. A 2-3 sentence summary
2. 3-5 key topics covered
3. Key takeaways (bullet points)

Transcript:
{transcript[:8000]}

Respond in JSON with keys: summary, topics, takeaways"""

    response = model.generate_content(prompt, generation_config=json_config)
    result = parse_json(response.text)
    if result:
        insight_cache.set(cache_key, result)
        return result
    fallback = {"summary": response.text, "topics": [], "takeaways": []}
    insight_cache.set(cache_key, fallback)
    return fallback


def summarize_playlist(videos: list[dict]) -> dict:
    video_summaries = "\n".join(
        f"- {v['title']}: {v['summary']}" for v in videos if v.get("summary")
    )

    prompt = f"""Given these videos from a YouTube playlist:
{video_summaries}

Provide:
1. A playlist overview (what this playlist is about)
2. Main themes across all videos
3. A suggested watch order with reasoning (which videos to watch first based on topic flow, each item should have title and reason)

Respond in JSON with keys: overview, themes, watch_order"""

    response = model.generate_content(prompt, generation_config=json_config)
    result = parse_json(response.text)
    if result:
        return result
    return {"overview": response.text, "themes": [], "watch_order": []}


def search_transcripts(query: str, transcripts: dict[str, str]) -> list[dict]:
    combined = "\n\n".join(
        f"[Video {vid}]: {text[:3000]}" for vid, text in transcripts.items() if text
    )

    prompt = f"""Search query: "{query}"

Find the most relevant sections from these video transcripts:
{combined[:12000]}

Return the top 3-5 most relevant results as JSON array with keys: video_id, relevance_snippet, score (0-100)"""

    response = model.generate_content(prompt, generation_config=json_config)
    result = parse_json(response.text)
    if isinstance(result, list):
        return result
    return []

File: api/intelligence.js
const { GoogleGenerativeAI } = require("@google/generative-ai");

function getModel() {
  const key = process.env.GEMINI_API_KEY;
  if (!key) throw new Error("GEMINI_API_KEY not set");
  const genAI = new GoogleGenerativeAI(key);
  return genAI.getGenerativeModel({
    model: "gemini-2.0-flash",
    generationConfig: {
      responseMimeType: "application/json",
      temperature: 0.3,
    },
  });
}

function parseJson(text) {
  try {
    return JSON.parse(text);
  } catch {
    const objMatch = text.match(/\{[\s\S]*\}/);
    if (objMatch) {
      try {
        return JSON.parse(objMatch[0]);
      } catch {}
    }
    const arrMatch = text.match(/\[[\s\S]*\]/);
    if (arrMatch) {
      try {
        return JSON.parse(arrMatch[0]);
      } catch {}
    }
    return null;
  }
}

function formatVideoEntry(v, i) {
  let entry = `${i + 1}. "${v.title}" (${v.video_id})`;
  if (v.transcript) {
    entry += `\nTranscript: ${v.transcript.slice(0, 1500)}`;
  } else if (v.description || v.tags?.length) {
    if (v.description) entry += `\nDescription: ${v.description.slice(0, 500)}`;
    if (v.tags?.length) entry += `\nTags: ${v.tags.join(", ")}`;
    if (v.duration) entry += `\nDuration: ${v.duration}`;
  }
  return entry;
}

async function analyzePlaylist(videos) {
  const model = getModel();

  const hasTranscripts = videos.filter((v) => v.transcript).length;
  const videoList = videos.map(formatVideoEntry).join("\n\n");

  const dataNote =
    hasTranscripts < videos.length
      ? `\nNote: ${hasTranscripts} of ${videos.length} videos have transcript data. For others, use the title, description, and tags to infer content. Use your knowledge of these videos and creators if you recognize them.`
      : "";

  const prompt = `You are analyzing a YouTube playlist. For each video, I have varying levels of data (transcript excerpts, descriptions, tags, or just titles). Provide a comprehensive analysis using all available information.${dataNote}

Videos:
${videoList.slice(0, 12000)}

Respond in JSON with these keys:
- overview: 2-3 sentences about what this playlist covers
- themes: array of 3-6 main themes across all videos
- videos: array of objects for each video with keys:
  - video_id: the video ID
  - summary: 1-2 sentence summary of what this specific video covers
  - topics: array of 2-4 topic tags
  - difficulty: "beginner", "intermediate", or "advanced" (if applicable, otherwise null)
  - prerequisites: which other videos in the playlist should be watched first (array of video_ids, empty if none)
- watch_order: array of objects with keys "video_id", "title", "reason" suggesting the best order to watch
- key_takeaways: array of 4-6 main things someone would learn from this entire playlist`;

  const result = await model.generateContent(prompt);
  const parsed = parseJson(result.response.text());
  if (!parsed) throw new Error("Failed to parse Gemini response");
  return parsed;
}

async function searchTranscripts(query, transcripts) {
  const model = getModel();

  const combined = Object.entries(transcripts)
    .filter(([, text]) => text)
    .map(([vid, text]) => `[Video ${vid}]: ${text.slice(0, 3000)}`)
    .join("\n\n")
    .slice(0, 12000);

  const prompt = `Search query: "${query}"

Find the most relevant sections from these video transcripts:
${combined}

Return the top 3-5 most relevant results as JSON array with keys: video_id, relevance_snippet, score (0-100)`;

  const result = await model.generateContent(prompt);
  const parsed = parseJson(result.response.text());
  return Array.isArray(parsed) ? parsed : [];
}

module.exports = { analyzePlaylist, searchTranscripts };
